compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1109,7 +1109,13 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
             // Encode this here because we don't do it in encode_def_ids.
             record!(self.tables.expn_that_defined[def_id] <- tcx.expn_that_defined(local_def_id));
         } else {
-            let direct_children = md.item_ids.iter().map(|item_id| item_id.def_id.local_def_index);
+            let direct_children = md.item_ids.iter().filter_map(|item_id| {
+                // Only encode named non-reexport children, reexports are encoded separately
+                // and unnamed items are not used by name resolution.
+                (tcx.def_key(item_id.def_id.to_def_id()).get_opt_name().is_some()
+                    && !matches!(tcx.hir().item(*item_id).kind, hir::ItemKind::ExternCrate(..)))
+                .then(|| item_id.def_id.local_def_index)
+            });
             // Foreign items are planted into their parent modules from name resolution point of view.
             let tcx = self.tcx;
             let foreign_item_children = md
```

Comment:
It's unfortunate that there's no way to easily make "heterogeneous iterators" for cases like this.

Unless you use a generator (https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Can.20generators.20be.20used.20in.20the.20compiler.3F).
Then it will look roughly like this:
```rust
record!(self.tables.children[def_id] <- gen_iter::gen_iter!({
	for item_id in md.item_ids {
		match tcx.hir().item(*item_id).kind {
			hir::ItemKind::ForeignMod { items, .. } => {
				for foreign_item in items {
					yield foreign_item.id.def_id.local_def_index;
				}
			}
			hir::ItemKind::ExternCrate(..) => continue,
			_ if tcx.def_key(item_id.def_id.to_def_id()).get_opt_name().is_some() => {
				yield item_id.def_id.local_def_index;
			}
			_ => continue,
		}
	}
}));
```

I'll probably add a `gen_iter`-like facility to `rustc_data_structures` tomorrow after all.
It will both look better and probably run faster as well.

Reply:
Custom iterator version.
(I don't want to collect into a vector, that's sort of admitting defeat).
A generator seems like a better alternative because it specifically exists to automatically generate the `enum` for you (including the flattening part).
```rust
            enum Iter<T1, T2, T3> {
                I1(T1),
                I2(T2),
                I3(T3),
            }
            impl<
                T1: Iterator<Item = DefIndex>,
                T2: Iterator<Item = DefIndex>,
                T3: Iterator<Item = DefIndex>,
            > Iterator for Iter<T1, T2, T3>
            {
                type Item = DefIndex;
                fn next(&mut self) -> Option<Self::Item> {
                    match self {
                        Iter::I1(i1) => i1.next(),
                        Iter::I2(i2) => i2.next(),
                        Iter::I3(i3) => i3.next(),
                    }
                }
            }

            record!(self.tables.children[def_id] <- {
                md.item_ids.iter().flat_map(|item_id| match tcx.hir().item(*item_id).kind {
                    hir::ItemKind::ForeignMod { items, .. } => {
                        Iter::I1(items.iter().map(|fi_ref| fi_ref.id.def_id.local_def_index))
                    }
                    hir::ItemKind::ExternCrate(..) => Iter::I2(iter::empty()),
                    _ if tcx.def_key(item_id.def_id.to_def_id()).get_opt_name().is_some() => {
                        Iter::I3(iter::once(item_id.def_id.local_def_index))
                    }
                    _ => Iter::I2(iter::empty()),
                })
            });
```